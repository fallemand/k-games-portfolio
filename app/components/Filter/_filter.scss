@import '../../styles/variables';
@import 'filter__small';

@media ( min-width: $breakpoint-desktop ) {
  @import 'filter__medium';
}

