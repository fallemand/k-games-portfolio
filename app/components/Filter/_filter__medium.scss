@import '../../styles/variables';

.filter {
  width: 250px;

  .filter__field {
    margin: 5px 0;
    border-radius: 4px 0 0 4px;
    -webkit-appearance: none;
    width: 40%;
    min-width: inherit;
  }

  .filter__value {
    margin: 5px 0 5px -1px;
    border-radius: 0 4px 4px 0;
    padding: 7px 10px 8px;
    -webkit-appearance: none;
    width: 60%;
    vertical-align: bottom;
  }
}
