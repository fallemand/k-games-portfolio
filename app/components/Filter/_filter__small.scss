@import '../../styles/variables';

.filter {
  width: 100%;

  .filter__field {
    width: 40%;
    display: inline-block;
  }

  .filter__value {
    border-radius: 0 4px 4px 0;
    border: 1px solid $border-color;
    padding: 8px 10px 7px;
    margin-left: -1px;
    width: 60%;
    display: inline-block;
  }
}
