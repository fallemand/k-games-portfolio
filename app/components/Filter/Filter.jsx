import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import Dropdown from '../Dropdown';
import './_filter.scss';

class Filter extends React.Component {
  constructor(props) {
    super(props);
    this.handleSortChange = this.handleSortChange.bind(this);
    this.handleRandomize = this.handleRandomize.bind(this);
    this.handleQueryChange = this.handleQueryChange.bind(this);
  }

  handleSortChange(value) {
    const { onChange, filter } = this.props;
    onChange({ sort: value, filter });
  }

  handleRandomize() {
    this.handleSortChange(Filter.sort.RANDOM);
  }

  handleQueryChange(event) {
    const { value } = event.target;
    const { onChange, sort } = this.props;
    onChange({ filter: value, sort });
  }

  render() {
    const { className, filter, sort } = this.props;
    const sortOrderFields = Object.entries(Filter.sort).map(
      ([key, value]) => ({ label: value, value }),
    );
    return (
      <form className={classnames(className, 'filter')}>
        <fieldset className="filter__fieldset">
          <legend className="filter__legend">Name</legend>
          <input
            name="filter"
            className="filter__field"
            placeholder="Game name"
            type="text"
            value={filter}
            onChange={this.handleQueryChange}
          />
        </fieldset>
        <fieldset className="filter__fieldset">
          <legend className="filter__legend">Sort Order</legend>
          <Dropdown
            name="sort"
            title="Sort"
            className="filter__field"
            value={sort}
            options={sortOrderFields}
            onChange={this.handleSortChange}
          />
        </fieldset>
        <fieldset className="filter__fieldset">
          <legend className="filter__legend">Randomize</legend>
          <button onClick={this.handleRandomize} type="button" className="filter__button">Go!</button>
        </fieldset>
      </form>
    );
  }
}

Filter.sort = {
  ASC: 'ascendant',
  DESC: 'descendant',
  RANDOM: 'random',
};

Filter.propTypes = {
  className: PropTypes.string,
  sort: PropTypes.oneOf(Object.values(Filter.sort).concat([''])),
  onChange: PropTypes.func.isRequired,
  filter: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

Filter.defaultProps = {
  className: null,
  filter: null,
  sort: null,
};

export default Filter;
