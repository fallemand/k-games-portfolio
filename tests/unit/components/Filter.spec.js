import React from 'react';
import Filter from '../../../app/components/Filter';

describe('Filter', () => {
  const props = {
    className: '__CLASSNAME__',
    sort: Filter.sort.ASC,
    onChange: jest.fn(),
    filter: '__QUERY__',
  };

  it('render with props', () => {
    const component = shallow(<Filter {...props} />);
    expect(escapeSnapshot(component)).toMatchSnapshot();
  });

  it('click on randomize should sort random', () => {
    const component = shallow(<Filter {...props} />);
    component.find('.filter__button').simulate('click');
    expect(props.onChange).toHaveBeenCalledWith({ sort: 'random', filter: '__QUERY__' });
  });

  it('change filter value should call onChange', () => {
    const component = shallow(<Filter {...props} />);
    component.find('input').simulate('change', { target: { value: '_test1_' } });
    expect(props.onChange).toHaveBeenCalledWith({ sort: 'ascendant', filter: '_test1_' });
  });

  it('select dropdown options should call onChange', () => {
    const component = shallow(<Filter {...props} />);
    component.find('Dropdown').simulate('change', '_test2_');
    expect(props.onChange).toHaveBeenCalledWith({ sort: '_test2_', filter: '__QUERY__' });
  });
});
