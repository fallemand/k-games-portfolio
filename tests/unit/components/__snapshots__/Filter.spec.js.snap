// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter render with props 1`] = `
"<form className='__CLASSNAME__ filter'>
  <fieldset className='filter__fieldset'>
    <legend className='filter__legend' id='name'>
      Name
    </legend>
    <input className='filter__field' placeholder='Game name' type='text' value='__QUERY__' onChange={[Function: bound handleQueryChange]} />
  </fieldset>
  <fieldset className='filter__fieldset'>
    <legend className='filter__legend' id='name'>
      Sort Order
    </legend>
    <Dropdown name='sort' title='Sort' className='filter__field' value='ascendant' options={{...}} onChange={[Function: bound handleSortChange]} />
  </fieldset>
  <fieldset className='filter__fieldset'>
    <legend className='filter__legend' id='name'>
      Randomize
    </legend>
    <button onClick={[Function: bound handleRandomize]} type='button' className='filter__button'>
      Go!
    </button>
  </fieldset>
</form>"
`;
